### ISSUE TITLE
ENV: Not functional on Firefox

### ISSUE THREAD

**Issue by alcen** [source-academy/cadet-frontend#705]
I ran the following code (pulled from an actual CS1101S tutorial sheet) on the Playground.

```
function memoize(f) {
    const mem = [];
    
    function mf(x) {
        if (mem[x] !== undefined) {
            return mem[x];
        } else {
            const result = f(x);
            mem[x] = result;
            return result;
        }
    }
    
    return mf;
}

const mfib = memoize(n => (n <= 1) ? n : mfib(n - 1) + mfib(n - 2));

mfib(3);
```

The breakpoint had been set to the last line. Although the debugger gave the correct values for the variables, the environmental model failed to show up at all.

<img width="960" alt="mfib" src="https://user-images.githubusercontent.com/49450743/61094894-bbcdab80-a483-11e9-84ba-7cc9f72972e0.png">
<img width="960" alt="mfib2" src="https://user-images.githubusercontent.com/49450743/61094897-c12af600-a483-11e9-982e-734d3db1e44c.png">

I tried this on the latest commit and the live version at https://sourceacademy.nus.edu.sg/playground.
Both times, the Debugger and Env Visualizer icons flashed orange. The correct variables and bindings appeared on the Debugger, but were nowhere to be seen on the Env Visualizer.

Also, is math_toSource supposed to be exposed in the Debugger/Env Visualizer?

**Comment by geshuming** [comment]
Hm, it's working for me...
![Capture](https://user-images.githubusercontent.com/43642522/61262082-8ba14800-a7b6-11e9-9515-f4f7f27102e1.PNG)

I'm not sure about the `math_toSource` function though

**Comment by geshuming** [comment]
@alcen you're using firefox right? It doesn't show up on firefox for me as well, but it does on chrome. The visualizer does not seem to work properly on firefox

**Comment by martin-henz** [comment]
> Hm, it's working for me...
> ![Capture](https://user-images.githubusercontent.com/43642522/61262082-8ba14800-a7b6-11e9-9515-f4f7f27102e1.PNG)
> 
> I'm not sure about the `math_toSource` function though

Do you also see the function math_toSource @geshuming ?

**Comment by alcen** [comment]
> @alcen you're using firefox right? It doesn't show up on firefox for me as well, but it does on chrome. The visualizer does not seem to work properly on firefox

Yeah, I was using Firefox. Might be a problem with the CSS then?

**Comment by alcen** [comment]
Seems like the Env Visualizer isn't entirely problem-free on Chrome either...
I tested this code on Google Chrome at https://sourceacademy.nus.edu.sg/playground:

```
const a = 5;
const b = "Hello world!";
const greetings = () => a + 1;

const f = 6;
a;
```

When you set the breakpoints to any of the first three lines the Env Visualizer works as expected.

![chrome74](https://user-images.githubusercontent.com/49450743/61263423-5b0fdd00-a7bb-11e9-86f1-e59dfec5760c.png)

&nbsp;
But if you refresh the page and set it to the 5th line (`const f = 6;`) the Env Visualizer fails to appear. Unless there is a hard-coded limit of 3 constants, this seems unintended

![chrome75](https://user-images.githubusercontent.com/49450743/61263427-606d2780-a7bb-11e9-96aa-0a241108f451.png)


@geshuming Could you try testing this on your system to confirm?

**Comment by alcen** [comment]
> Hm, it's working for me...
> 
> I'm not sure about the math_toSource function though
> 
> Do you also see the function math_toSource @geshuming ?

Seems like math_toSource is not exposed on Chrome.

**Comment by martin-henz** [comment]
The issue with math_toSource will go away after this is handled: https://github.com/source-academy/js-slang/issues/255

**Comment by martin-henz** [comment]
@lumos309 can you take a look at the ENV issues?

**Comment by openorclose** [comment]
> The issue with math_toSource will go away after this is handled: [source-academy/js-slang#255](https://github.com/source-academy/js-slang/issues/255)

It won't,  because the inspector manually hides all the builtins. Firefox has Math.toSource but Chrome doesn't so the manual hiding still needs to be done.

**Comment by lumos309** [comment]
> Seems like the Env Visualizer isn't entirely problem-free on Chrome either...
> ...
> When you set the breakpoints to any of the first three lines the Env Visualizer works as expected.
> ...
> But if you refresh the page and set it to the 5th line (`const f = 6;`) the Env Visualizer fails to appear. Unless there is a hard-coded limit of 3 constants, this seems unintended

It's actually the arrow function that breaks the visualiser, not a limit on the number of names. I have a fix along with some other updates for the visualiser coming soon.

I'll take a look at the Firefox-only issues.

**Comment by martin-henz** [comment]
> > Seems like the Env Visualizer isn't entirely problem-free on Chrome either...
> > ...
> > When you set the breakpoints to any of the first three lines the Env Visualizer works as expected.
> > ...
> > But if you refresh the page and set it to the 5th line (`const f = 6;`) the Env Visualizer fails to appear. Unless there is a hard-coded limit of 3 constants, this seems unintended
> 
> It's actually the arrow function that breaks the visualiser, not a limit on the number of names. I have a fix along with some other updates for the visualiser coming soon.
> 
> I'll take a look at the Firefox-only issues.

OK. Do you have a timeline in mind? PR when? Reviewer: who?  I'm trying to coordinate the deployment over the next week or so. Tx.

**Comment by lumos309** [comment]
The Firefox issue is actually caused by the `math_toSource` built-in function. Both the inspector and the env visualizer have a list of built-in functions that should be ignored; with `math_toSource` not being exposed on Chrome neither team found it and included it in this list and so it broke the visualizer and incorrectly shows up in the inspector.

In fact, there appears to be a new issue when any library (RUNES, CURVES etc.) is selected, where the additional built-in functions from these libraries also breaks the visualizer and causes a bunch of "behind the scenes" code to appear in the inspector. This should also be easily fixed by manually adding these functions to the above built-in list. @martin-henz @podocarp

**Comment by lumos309** [comment]
> OK. Do you have a timeline in mind? PR when? Reviewer: who? I'm trying to coordinate the deployment over the next week or so. Tx. @lumos309

Addressed in #800 @martin-henz .

**Comment by alcen** [comment]
Resolved in #800 'Various Bug Fixes'!!
